src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,164 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> tail;
+    private Node<T> head;
+    private int size;
+
     @Override
     public void add(T value) {
+        Node<T> node = new Node<>(value);
+        if (size == 0) {
+            head = tail = node;
+        } else {
+            tail.next = node;
+            node.prev = tail;
+            tail = node;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        checkAddedIndex(index);
+        Node<T> newNode = new Node<>(value);
+        if (head == null) {
+            head = tail = newNode;
+        } else if (index == 0) {
+            newNode.next = head;
+            head.prev = newNode;
+            head = newNode;
+        } else if (index == size) {
+            newNode.prev = tail;
+            tail.next = newNode;
+            tail = newNode;
+        } else {
+            Node<T> current = getNodeByIndex(index);
+            newNode.next = current;
+            newNode.prev = current.prev;
+            current.prev.next = newNode;
+            current.prev = newNode;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        outOfBoundsCheck(index);
+        Node<T> current = getNodeByIndex(index);
+        return (T) current.value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        outOfBoundsCheck(index);
+        Node<T> current = getNodeByIndex(index);
+        T removedValue = current.value;
+        current.value = value;
+        return removedValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        outOfBoundsCheck(index);
+        if (index == 0) {
```

Comment:
do we really need it if we just delink the element obtained by the method getNodeByIndex().value;? 
This will be more optimal, in my opinion, but the unlink method in this case, also needs to be improved

Reply:
agree. Let’s fix it ;)